tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/**
+ * OAI-PMH harvester factory integration test.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2010.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+namespace VuFindTest\Harvest\OaiPmh;
+
+use Laminas\Http\Client;
+use VuFindHarvest\OaiPmh\Communicator;
+
+/**
+ * OAI-PMH communicator test.
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+class CommunicatorTest extends \PHPUnit\Framework\TestCase
+{
+
+    /**
+     * Get Communicator
+     *
+     * @param string                     $uri       Base URI for OAI-PMH server
+     * @param Client                     $client    HTTP client
+     *
+     * @return type
+     */
+    protected function getCommunicator($uri, $client)
+    {
+        return new Communicator($uri, $client);
```

Comment:
It might make sense to add the ability to inject the Communicator with a mock \VuFindHarvest\ConsoleOutput\OutputInterface (using `setOutputWriter()`) so that you can set expectations about output ... e.g. to capture the logic that displays a "wait" message during the retry loop.

Reply:
Hummm, are you suggesting to do this inside getCommunicator(), but then access the mocked writer later in each test? It's not clear to me how I could do that in a generic way, but I may certainly be missing something simple about your suggestion. All that said, in https://github.com/vufind-org/vufindharvest/pull/10/commits/3e8884b82ce2b859c6ed422e14cdb6692ce5c96a I did add a test-specific version of this to check that the Console output line about the 503 response was written.

Reply:
There are a few ways you could approach this (e.g. create an $outputWriter property in the test class, and then set that to a mock writer in getCommunicator; or, add a second parameter to getCommunicator so you could preconfigure your output writer and inject it), but none of them are especially satisfying or clean-feeling, so I think addressing this on a test-by-test basis as you have makes sense here.